palette/src/oklab.rs

```diff
@@ -0,0 +1,786 @@
+use core::marker::PhantomData;
+use core::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Sub, SubAssign};
+
+#[cfg(feature = "random")]
+use rand::distributions::uniform::{SampleBorrow, SampleUniform, Uniform, UniformSampler};
+#[cfg(feature = "random")]
+use rand::distributions::{Distribution, Standard};
+#[cfg(feature = "random")]
+use rand::Rng;
+
+use crate::color_difference::ColorDifference;
+use crate::color_difference::{get_ciede_difference, LabColorDiff};
+use crate::convert::FromColorUnclamped;
+use crate::encoding::pixel::RawPixel;
+use crate::matrix::multiply_xyz;
+use crate::white_point::{WhitePoint, D65};
+use crate::{
+    clamp, contrast_ratio, from_f64, Alpha, Component, ComponentWise, FloatComponent, GetHue,
+    Limited, Mat3, Mix, OklabHue, Pixel, RelativeContrast, Shade, Xyz,
+};
+
+#[rustfmt::skip]
+fn m1<T: FloatComponent>() -> Mat3<T> {
+    [
+        from_f64(0.8189330101), from_f64(0.3618667424), from_f64(-0.1288597137),
+        from_f64(0.0329845436), from_f64(0.9293118715), from_f64(0.0361456387),
+        from_f64(0.0482003018), from_f64(0.2643662691), from_f64(0.6338517070),
+    ]
+}
+
+#[rustfmt::skip]
+pub(crate) fn m1_inv<T: FloatComponent>() -> Mat3<T> {
+    [
+        from_f64(1.2270138511), from_f64(-0.5577999807), from_f64(0.2812561490),
+        from_f64(-0.0405801784), from_f64(1.1122568696), from_f64(-0.0716766787),
+        from_f64(-0.0763812845), from_f64(-0.4214819784), from_f64(1.5861632204),
+    ]
+}
+
+#[rustfmt::skip]
+fn m2<T: FloatComponent>() -> Mat3<T> {
+    [
+        from_f64(0.2104542553), from_f64(0.7936177850), from_f64(-0.0040720468),
+        from_f64(1.9779984951), from_f64(-2.4285922050), from_f64(0.4505937099),
+        from_f64(0.0259040371), from_f64(0.7827717662), from_f64(-0.8086757660),
+    ]
+}
+
+#[rustfmt::skip]
+pub(crate) fn m2_inv<T: FloatComponent>() -> Mat3<T> {
+    [
+        from_f64(0.9999999985), from_f64(0.3963377922), from_f64(0.2158037581),
+        from_f64(1.0000000089), from_f64(-0.1055613423), from_f64(-0.0638541748),
+        from_f64(1.0000000547), from_f64(-0.0894841821), from_f64(-1.2914855379),
+    ]
+}
+
+/// Oklab with an alpha component. See the [`Oklaba` implementation in
+/// `Alpha`](crate::Alpha#Oklaba).
+pub type Oklaba<Wp = D65, T = f32> = Alpha<Oklab<Wp, T>, T>;
+
+/// The [Oklab color space](https://bottosson.github.io/posts/oklab/).
+///
+/// Oklab is a perceptually-uniform color space similar in structure to [L\*a\*b\*](crate::Lab), but
+/// with better perceptual uniformity. It assumes a D65 whitepoint and normal well-lit viewing
+/// conditions. As such, **the `Wp` type parameter is ignored** and is only present due to
+/// complications with `FromColorUnclamped`.
+#[derive(Debug, PartialEq, Pixel, FromColorUnclamped, WithAlpha)]
+#[cfg_attr(feature = "serializing", derive(Serialize, Deserialize))]
+#[palette(
+    palette_internal,
+    white_point = "Wp",
```

Comment:
I changed `Wp` to `D65` and removed the `Wp` type parameter, and got an error like this

```console
error[E0107]: wrong number of type arguments: expected at most 1, found 2
  --> palette/src/rgb/rgb.rs:46:35
   |
46 | #[derive(Debug, PartialEq, Pixel, FromColorUnclamped, WithAlpha)]
   |                                   ^^^^^^^^^^^^^^^^^^ expected at most 1 type argument
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

at every usage site of the `FromColorUnclamped` derive macro.

Reply:
Should I push the ‘more correct’ but broken version that doesn’t have the `Wp` parameter now, or should I wait until we fix all the errors?

Reply:
My guess is that it expands to a give it a white point it shouldn't have, when implementing it as a part of the default set. There are a few places in the macro code where I make a similar exception for the RGB group of types, so you can program it to have the correct parameters there.

If you push your work in progress, I can help nore and direct you along the way where needed. 🙂 It's always possible to tidy up the commits afterwards.

Reply:
I tried my best to fix the issue in eefb47e, but got a different error now about `FromColorUnclamped` not being implemented when converting from an `Xyz` of an arbitrary colour space to an `Oklab`, which only has an implementation for `D65`. Not sure where to go from here 😕

Reply:
Is this in the other FromColorUnclamped implementations? If not, it's totally expected and desired. It shouldn't switch white point automatically. But I assume that's not the case. We'll have to restrict the white point to D65 all over the place if the target type is Oklab, but it looks like that may require a bit more work than I thought. But I'm leaning towards adding `#nearest_color_path<#white_point, #component>: #from_trait_path<#color_ty>` in the where clause for the Oklab case. I can give it a try and see if it works.

Reply:
So, I think I managed to fix it. Adding the where clause did the trick, since it basically says that we assume it will work, even when it doesn't. Here's what I arrived at after poking around a bit: https://github.com/Ogeon/palette/commit/6014fc9643693a410f68b777bc26869a6a008103

It would probably be nicer to add a list for special requirement types like this, so the `if` condition doesn't grow out of proportion, but I decided to keep this small.

Reply:
For example, here's what it generates for `Lab`:

````rust
impl<Wp, T> FromColorUnclamped<Oklab<T>> for Lab<Wp, T> where
    T: FloatComponent,
    Wp: WhitePoint,
    T: FloatComponent,
    Wp: WhitePoint,
    T: FloatComponent,
    Xyz<Wp, T>: FromColorUnclamped<Oklab<T>>, // <--------
````

Reply:
That’s pretty cool! How can I rebase your commit onto my branch?

Reply:
I think you can use `git pull`, the same way I did to get your branch: `git pull git://github.com/Ogeon/palette.git arzg-add-oklab`. Make sure you are in your `add-oklab` when doing it. If that doesn't work, just copy the file content. I don't need my name on it. 😁

Reply:
Done, thanks!